source/_posts/2025-07-18-why-i-disallow-facade-usage.md

```diff
@@ -0,0 +1,333 @@
+</p>
+
+<p>
+    @TODO expose concept of "simple" != "easy". Systems can be complex and easy, or harder to use, but simple.
+</p>
+
+<h2>The simple solution</h2>
+
+<p>
+    Here's our service re-implemented to make things simple and explicit:
+</p>
+
+~~~php
+<?php
+
+final class MyService implements SomeContract {
+    public function __construct(private readonly SomeCache) {}
+    public function doSomeHeavyWork(): Result
+    {
+        $result = $this->heavyLifting();
+
+        $this->cache->put('a key', $result);
+
+        return $result;
+    }
+
+    private function heavyLifting(): Result
+    {
+        // ... irrelevant ...
+    }
+}
+~~~
+
+<p>
+    The above is a little more verbose, but follows very old and well-functioning
+    <a href="https://www.youtube.com/watch?v=RlfLCWKxHJ0">dependency injection</a> rules.
+</p>
+
+
+<p>
+    By adding 2 lines of code, we:
+</p>
+
+<ul>
+    <li>make the shared state visible</li>
+    <li>removes a framework dependency</li>
+    <li>removes runtime crashes in case of misconfiguration</li>
+    <li>allow multiple instances of a service to use different dependencies</li>
+    <li>makes the dependency visible at reflection level (useful for dependency analysis tooling)</li>
+</ul>
+
+<p>
+    Does the above solution have worse <abbr title="Developer Experience">DX</abbr>?
+    Given all the listed advantages, I think it provides much better experience.
+</p>
+
+<p>
+    You have to also remember that facades were designed and built in an age when auto-wiring dependency injection
+    containers weren't common in PHP: the simpler approach may actually even be easier, when using the full framework.
+</p>
+
+<h2>But what about my test helpers?</h2>
+
+<p>
+    @TODO Laravel has MyFacade::spy() and MyFacade::mock() helpers - let's document those
+</p>
+
+~~~php
+<?php
+
+final class MyServiceTest extends \Illuminate\Foundation\Testing\TestCase {
+    function test_something_is_being_done(): void
+    {
+        Cache::spy();
+        
+        Cache::shouldReceive('put')
+            ->with('a key', 'a value');
+        
+        $service = new MyService();
+        
+        $result = $service->doSomeHeavyWork();
+        
+        // more assertions on $result
+    }
+}
+~~~
+
+<p>
+    The above example swaps the facade underlying service location mechanism with a mock, and relies on an
+    <a href="https://github.com/laravel/framework/blob/2c682e4eb531eae6439579f25d29429ecc0a66ca/src/Illuminate/Foundation/Testing/TestCase.php#L9-L21">
+        extremely complex base test class
+    </a> to do framework startup/cleanup operations.
+</p>
+
+<p>
+    Here's the "simple" version instead:
+</p>
+
+
+~~~php
+<?php
+
+final class MyServiceTest extends \PHPUnit\Framework\TestCase {
+    function test_something_is_being_done(): void
+    {
+        $cache = $this->createMock(Cache::class);
```

Comment:
Not sure if teaching Mocks is the best thing to do? I'd ask if a `Spy` would be smarter to use here?

Reply:
Sorry, missed the next part -_-